src/garrysmod/gamemodes/zombiesurvival/gamemode/sv_mapvote.lua

```diff
@@ -0,0 +1,202 @@
+    query_obj:set_sync(true)
+    local success, res, size = query_obj:run()
+    if not success then
+        ErrorNoHalt(res)
+    end
+
+    -- If there are any maps present in the database that are *not* present on
+    -- the server, let's update their status
+    base = [[
+        UPDATE
+            maps
+        SET
+            installed = FALSE
+        WHERE
+            id NOT IN (%s);
+    ]]
+
+    values = {}
+    for i=1,#maps do
+        table.insert(values, string.format("%s", conn:quote(conn:escape(maps[i]))))
+    end
+    query_obj = conn:query(string.format(base, table.concat(values, ", ")))
+    query_obj:set_sync(true)
+    local success, res, size = query_obj:run()
+    if not success then
+        ErrorNoHalt(res)
+    end
+end
+
+function GM.MapVote:MakePool()
+    local conn = GAMEMODE.Database.conn
+    local base = [[
+        SELECT
+            *
+        FROM
+            maps
+        WHERE
+            installed = TRUE
+            AND id != %s
+            AND last_played < (now() - interval '%d seconds')
+        ORDER BY
+            random()
+        LIMIT 8;
+    ]]
+    local query_obj = conn:query(string.format(base, conn:quote(conn:escape(game.GetMap())), GAMEMODE.MapVote.cooldown))
+    query_obj:set_sync(true)
+    local success, res, size = query_obj:run()
+    if success then
+        local pool = {}
+
+        for i=1,#res do
+            table.insert(pool, res[i].id)
+        end
+
+        return pool
+    else
+        ErrorNoHalt(res)
+    end
+end
+
+function GM.MapVote:NetworkToPlayer(player)
+    if self.Active then
+        net.Start("zs_mapvote_start")
+        net.WriteUInt(self.EndsAt, 16)
+        net.WriteTable(self.Pool)
+        net.Send(player)
+    end
+end
+
+function GM.MapVote:Start(duration) 
+    if self.Active then
+        error("MapVote is already active.")
+    end
+
+    local duration = duration or 30
+
+    self.Active = true
+    self.EndsAt = CurTime() + duration
+    self.Pool = self:MakePool()
+    self.Votes = {}
+
+    net.Start("zs_mapvote_start")
+    net.WriteUInt(self.EndsAt, 16)
+    net.WriteTable(self.Pool)
+    net.Broadcast()
+  
+    timer.Simple(duration, function()
+        self:End()
+    end)
+end
+
+function GM.MapVote:Tally()
+    if table.IsEmpty(self.Pool) then
+        return nil
+    elseif table.IsEmpty(self.Votes) then
+        MsgN("No one voted. MapVote tally is random.")
+        return math.random(1, #self.Pool)
+    end
+
+    local count = {}
+
+    for _,vote in pairs(self.Votes) do
+        count[vote] = count[vote] or 0 + 1
+    end
+
+    local winner = table.GetWinningKey(count)
+
+    return winner
+end
+
+function GM.MapVote:End()
+    if not self.Active then
+        error("MapVote is already over.")
+    end
+
+    self.Active = false
+
+    local tally = self:Tally()
+
+    if not tally then
+        error("MapVote tally cannot be nil. Aborting.")
```

Comment:
Although this shouldn't ever happen, if it does, then it looks to me like the server will get stuck on the current map, which doesn't sound like desired behaviour.

Reply:
This comment made me spend a couple hours working on 689ebec 😞. I think the connection to the database is much more resilient now. We have a queue for queries we can run asynchronously, and we have a function for determining whether or not the connection is valid.

Now, this should be much less likely to occur, because we can fallback to a pool of maps solely from the filesystem.

I think we should still leave this in. If it gets to this point, they're running a server with no maps installed, and I'd like things to blow up in a kind way.